docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,883 @@
+
+Cluster API is going to implement calls to registered Runtime Extensions at well-known moments of the Cluster’s lifecycle.
+
+The two key elements that make the implementation of runtime extension calls simple and consistent across
+the codebase are:
+
+- The Runtime Hook catalog, providing the info about all the defined Runtime Hooks, supported version and
+  corresponding input/output types;
+- The Runtime Extensions registry, providing info about the registered Runtime Extensions implementing the
+  Runtime Hooks defined above.
+
+Given these two elements, the code for calling Runtime Extensions is:
+
+_Note: this is pseudo code, meant to demonstrate a set of elements described below the example; the exact details
+will be defined during implementation, possibly taking advantage of golang generics._
+
+```go
+extensions := registry.Get(
+    registry.Group("cluster.runtime.cluster.x-k8s.io/v1alpha2”),
+    registry.Hook(&v1alpha2.BeforeUgradeHook{}),
+)
+
+for _, e := range extensions{
+    client := catalogHTTP.NewClientBuilder().
+        WithCatalog(c).
+        Host(e.host).
+        Build()
+
+    hook := &v1alpha2.BeforeUgradeHook{}
+    in := &v1alpha2.BeforeUgradeInput{First: 1, Second: "Hello CAPI Runtime Extensions!"}
+    out := &v1alpha2.BeforeUgradeOutput{}
+    if err := client.Extension(hook, catalogHTTP.SpecVersion(r.version)).Invoke(ctx, in, out); err != nil {
+        panic(err)
+    }
+
+    // Do something with the output e.g. proceed with upgrade or block
+}
+```
+
+A couple of elements are worth noting:
+
+- Registered Runtime Extensions are returned by group and hook; this will also include Runtime Extensions
+  implementing older versions of the same Runtime Hook;
+- The call is implemented using the last version of the Runtime Hook/Input/Output types; the Invoke function
+  will take care of version conversions, if required.
+
+## Security Model
+
+Following threats were considered:
+
+- Malicious Runtime Extensions being registered 
+
+Mitigation: The same mitigations used for avoiding malicious dynamic webhooks in Kubernetes apply
+(defining RBAC rules for the RuntimeExtensionConfiguration assigning this responsibility to cluster admin only).
+
+- Privilege escalation of HTTP Servers running Runtime Extensions
+
+Mitigation: The same mitigations used for any HTTP server deployed in Kubernetes apply
+(use distroless base image, do not use privileged pods etc.).
+
+- Tampering of the communication channel between Cluster API controllers and HTTP Servers implementing Runtime Extensions.
+
+Mitigation: The same mitigations used for any HTTP server deployed in Kubernetes apply (use SSL, Network policies etc.).
+
+## Risks and Mitigations
+
+- Building Runtime SDK, Runtime Hooks and Runtime Extensions in sequential steps might lead to reworks.
+
+This is an accepted risk, given the importance of defining a robust SDK before external developers start relying
+extensively on this feature.
+
+## Alternatives
```

Comment:
Perhaps, another alternative would be to use a WASM module. A WASM engine could be hosted in process and load a module to handle specific events. There would be no need for grpc based plugins since the WASM module could run sandboxed in-process.

Prior art:
- https://www.openpolicyagent.org/docs/latest/wasm/
- https://istio.io/latest/docs/reference/config/proxy_extensions/wasm-plugin/
- https://docs.kubewarden.io

Reply:
this alternative is already documented in "Using in-process plugins vs calling external components". I have also experimented on it but then discarded it for the reason described below; Personally, I also find that the golang support for web assembly seems very limiting, and introducing another language/toolchain is not practical.

Reply:
My fault. I missed the WebAssembly call out.

The trade off seems to be operational / perf pain from external dependencies vs. dev / tool chain pain. Personally, the latter would be more interesting to me as that space is quickly improving where the ops / perf of using external services would be forever.

Reply:
Closing as per March 23th office hours discussion